orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/OptionalStageSupport.groovy

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.netflix.spinnaker.orca.pipeline.model
+
+import com.netflix.spinnaker.orca.pipeline.util.ContextParameterProcessor
+
+class OptionalStageSupport {
+  public static Map<String, Class<? extends OptionalStageEvaluator>> OPTIONAL_STAGE_TYPES = [
+    "parameter" : ParameterOptionalStageEvaluator,
+    "expression": ExpressionOptionalStageEvaluator
+  ]
+
+  /**
+   * A Stage is optional if it has an {@link OptionalStageEvaluator} in it's context that evaluates {@code true}.
+   */
+  static boolean isOptional(Stage stage) {
+    def optionalType = (stage.context.optional?.type as String)?.toLowerCase()
+    if (!optionalType || !OPTIONAL_STAGE_TYPES[optionalType]) {
+      return false
+    }
+
+    return stage.mapTo("/optional", OPTIONAL_STAGE_TYPES[optionalType]).isOptional(stage)
+  }
+
+  /**
+   * Determines whether a stage is optional and should be skipped
+   */
+  private static interface OptionalStageEvaluator {
+    boolean isOptional(Stage stage)
+  }
+
+  /**
+   * An {@link OptionalStageEvaluator} that will perform a case-insensitive comparison against a pipeline parameter.
+   */
+  private static class ParameterOptionalStageEvaluator implements OptionalStageEvaluator {
```

Comment:
thoughts on special casing this instead of just using the equivalent expression? `parameters.key == value` as the expression doesn't seem much worse and saves having an extra case both here and in deck